include/envoy/server/filter_config.h

```diff
@@ -200,6 +200,14 @@ class FactoryContext : public virtual CommonFactoryContext {
   virtual ProtobufMessage::ValidationVisitor& messageValidationVisitor() PURE;
 };
 
+/**
+ * An implementation of FactoryContext. It should be used to create NetworkFilterChain.
+ */
+class FilterChainFactoryContext : public virtual FactoryContext {};
```

Comment:
Why do we need this at the interface level? Do you expect to add more methods later? The implementation of the interface can of course be different, but we should not add a new interface unless we really need it?

Reply:
We now have two impl of FactoryContext:
`ListenerFactoryContext` and `FilterChainFactoryContext`

In this PR, where ever the function needs a `FilterChainFactoryContext`, if you pass a `ListenerFactoryContext` instead, the envoy could die. 

However, if we eliminate `FilterChainFactoryContext` interface here, I could not provide the concept: I need a `FactoryContext` but please don't give me a `ListenerFactoryContext`

To me it's critical because I feel lifetime should be part of the interface. But if most people have the knowledge of where `ListenerFactoryContext` is not `FactoryContext`, this `FilterChainFactoryContext` is indeed an overhead

Reply:
That's fine if you want to keep it for type checking, but I would make the comment clear on why we are doing this.

Reply:
got it. Will update the comment adding life cycle description